src/stan/services/sample/hmc_nuts_diag_e_adapt.hpp

```diff
@@ -53,20 +53,20 @@ namespace sample {
  * @param[in,out] diagnostic_writer Writer for diagnostic information
  * @return error_codes::OK if successful
  */
-template <class Model>
+template <typename Model, typename InitContext, typename InitMetricContext,
+          typename SampleWriter, typename DiagnosticWriter>
 int hmc_nuts_diag_e_adapt(
-    Model& model, const stan::io::var_context& init,
-    const stan::io::var_context& init_inv_metric, unsigned int random_seed,
+    Model& model, const InitContext& init,
+    const InitMetricContext& init_inv_metric, unsigned int random_seed,
```

Comment:
Are the types changing here? I don't see an advantage to templating this. Now the template hides even the base polymorphic type.

Reply:
Ah they are! Notice that now when an initial inverse metric is not given we actually return back an `stan::io::dump` (or vector of). This is nice for the compiler because now it knows the real type coming in so no devirtualization happens a lot easier.

Reply:
Is a `var_context` more informative? It's just an abstract base class so you still need to go look at the callee to see what's happening when this objects member functions are called. What if I included in the docs that this is either going to be a `stan::io::dump` or `stan::io::json_context` etc.?

Reply:
> Is a var_context more informative

Well it's still easier to search for (you can find that it's an abstract base class, and then you can find what inherits from it and whatnot) and this is how everything is kinda done in the rest of services.

I think the reason we'd go templating here is if we were doing something we couldn't just solve with the existing polymorphism, or it's just easier to code the templated thing.

If that's the case, let's do templating. If that's not the case, let's do templating it if you really want to. Not the biggest deal but I prefer the existing stuff.